File: frontend/src/api/gcps.ts
/**
 * ★★ Ground Control Points — endpoints 38–42 and 64 (§7), **plus the manual-mode
 *    create that §7 does not have**. THE DELIVERABLE.
 *
 * ─────────────────────────────────────────────────────────────────────────────
 * ★★ THE GAP, AND WHY THIS MODULE FILLS IT.
 *
 *   §7 has **no `POST` that creates a GCP**. Endpoints 38/39 read, 40 patches, 41
 *   resets, 42 recomputes, 64 lists by match result — every one of them presupposes
 *   that GCPs come into existence *as a side effect of a match job*. That was true of
 *   the system CONTRACT.md describes.
 *
 *   **SCOPE.md overrides it.** The matching engine is deferred, and manual GCP mode
 *   — *"photo pixel ↔ map click → lat/lon"* — is *"now the product's core
 *   interaction"* (SCOPE.md §5). With no create endpoint the product cannot produce
 *   a single GCP, which makes every other endpoint in this file a reader of an empty
 *   table. So `POST /images/{image_id}/gcps` is **added**, in the spirit of §7's
 *   nesting (`gcps.py` `router_image`, exactly where endpoints 38 and 42 already
 *   live), and flagged in the IU-24 report for IU-17/IU-19/IU-21.
 *
 *   {@link GcpCreate} is likewise declared here because IU-23 owns `types/**` and
 *   §8.2's `gcp.ts` row has no `GcpCreate` — for the same reason: the contract had no
 *   create. **It belongs in `types/gcp.ts`.** Flagged.
 * ─────────────────────────────────────────────────────────────────────────────
 */

import type { Page, Uuid } from '../types/common';
import type { GeoJsonFeatureCollection, PixelXY, ProviderId } from '../types/geo';
import type {
  GcpFilters,
  GcpOverview,
  GcpOverviewFilters,
  GcpRead,
  GcpRecomputeDryRun,
  GcpRecomputeRequest,
  GcpResetRequest,
  GcpUpdate,
  SurveyorConfidence,
} from '../types/gcp';
import type { JobRead } from '../types/job';
import { fetchJson, forgetEtag, toQuery } from './client';

// ─────────────────────────────────────────────────────────────────────────────
// The manual correspondence — SCOPE.md §5
// ─────────────────────────────────────────────────────────────────────────────

/**
 * ★★ `POST /images/{image_id}/gcps` — commit a manual correspondence.
 *
 * The surveyor marked a landmark in the photo and clicked the same physical spot on
 * the satellite map. **The coordinate is a DIRECT OBSERVATION, not an inference**
 * (SCOPE.md §2) — there is no homography, so there is no residual, no degeneracy
 * gate, and no computed confidence.
 *
 * Mirrors `correspondenceStore`'s committable state (IU-25) field for field, so the
 * commit is a projection and not a translation.
 */
export interface GcpCreate {
  /**
   * ★ **ORIGINAL image pixel space** — independent of viewer zoom, pan, brightness
   *   and contrast. Produced by `lib/viewport/transform.ts`'s `stageToImage` at
   *   `ImageViewer`'s single conversion site (§8.6: *"normative and
   *   correctness-critical"*). **Never rounded** — sub-pixel is stored, rounding is
   *   for display only.
   */
  /**
   * ★ Send **exactly one** of `image_px` or `landmark_id` — never both. The server
   *   rejects both with 422 ("Send exactly one … sending both makes two claims about one
   *   fact"): `landmark_id` derives the photo pixel from a saved annotation, `image_px`
   *   states it directly. When the correspondence is anchored to a saved landmark, send
   *   `landmark_id` and omit `image_px`; otherwise send `image_px`.
   */
  image_px?: PixelXY;

  /** ★ The map click, **EPSG:4326**. Stored as `geography(Point, 4326)` (SCOPE.md §5). */
  lat: number;
  lon: number;

  /**
   * ★ **SURVEYOR-DECLARED, NEVER COMPUTED** (SCOPE.md §5), and therefore **REQUIRED**.
   *
   *   Optional-with-a-default would manufacture a judgement the human never made,
   *   which is the same failure §8.7's decimal truncation exists to prevent, one
   *   field over. `correspondenceStore` refuses to reach `ready` without it for the
   *   same reason.
   */
  declared_confidence: SurveyorConfidence;

  /**
   * The annotation this correspondence is anchored to — `GcpRead.landmark_id`.
   * `null` when the surveyor clicked a bare pixel rather than an existing landmark.
   *
   * ★ `correspondenceStore` holds a `landmark_client_id` (annotationStore's local
   *   identity). The caller resolves it to the persisted annotation's `Uuid` before
   *   committing; an unsaved landmark has no server id to link.
   */
  landmark_id?: Uuid | null;

  /** `'GCP01'`. `^[A-Za-z0-9_\-]{1,32}$`, unique per image. Server-assigned when omitted. */
  code?: string | null;
  /** ★ The surveyor's free-text point name — the "Name" column (no character constraints). */
  name?: string | null;

  /**
   * ★ ACCURACY INPUTS, and they are not optional in spirit.
   *
   *   SCOPE.md §5: *"Reported positional accuracy comes from imagery
   *   ground-sample-distance and click precision at the map's zoom level — a real,
   *   defensible number — not from a match score."* The server cannot compute either
   *   term without knowing **which basemap** the click landed on and **at what
   *   zoom**: GSD is `f(provider, zoom, latitude)` and click precision is
   *   `f(zoom)`. Omit them and `gis/accuracy.py` has nothing to work from, leaving
   *   `total_ce90_m` — which §11.7 calls mandatory on every GCP — uncomputable.
   *
   *   Typed optional only because the server may fall back to the project's default
   *   provider; `map_zoom` has no such fallback and callers must send it.
   */
  provider?: ProviderId | null;
  map_zoom?: number | null;

  /**
   * The surveyor's free-text note. ★ The create wire field is `note` (the backend
   * `GcpManualCreate` schema; `extra="forbid"` 422s anything else — `adjustment_note`
   * included). It round-trips as `GcpRead.adjustment_note` on reads.
   */
  note?: string | null;
}

// ─────────────────────────────────────────────────────────────────────────────

const forImage = (imageId: Uuid): string => `/images/${imageId}/gcps`;
const flat = (id: Uuid): string => `/gcps/${id}`;

/** Endpoint 42's two outcomes: `dry_run: true` → `200`, otherwise `202 JobRead`. */
export type GcpRecomputeResponse = JobRead | GcpRecomputeDryRun;

/** Discriminates {@link GcpRecomputeResponse} structurally — a `JobRead` has a `status`. */
export function isRecomputeDryRun(r: GcpRecomputeResponse): r is GcpRecomputeDryRun {
  return typeof r === 'object' && r !== null && 'deltas' in r && 'fit' in r;
}

export const gcpsApi = {
  /**
   * ★★ ADDED — `POST /images/{image_id}/gcps` → `201 GcpRead`. Not in §7. See the
   *    module header. **This is how every GCP in this build comes into existence.**
   *
   * ★ The server sets `source: 'manual'` — it is not a client input. A client that
   *   could assert its own provenance could assert `'automatic'`, and SCOPE.md §5's
   *   whole point is that *"an automatic GCP can never be confused with an observed
   *   one downstream or in an export"*. Provenance is the server's to record.
   */
  create: (imageId: Uuid, body: GcpCreate, signal?: AbortSignal): Promise<GcpRead> =>
    fetchJson<GcpRead>(forImage(imageId), { method: 'POST', body, signal }),

  /**
   * `POST /images/{id}/gcps/copy-from/{sourceImageId}` → `201`. A camera that adopts
   * a NEW frame with the SAME aim carries the previous frame's control points over
   * (2026-09-09) — each recreated through the manual path, same pixel, same spot.
   */
  copyFrom: (
    imageId: Uuid,
    sourceImageId: Uuid,
    signal?: AbortSignal,
  ): Promise<{ copied: number; items: GcpRead[] }> =>
    fetchJson<{ copied: number; items: GcpRead[] }>(
      `${forImage(imageId)}/copy-from/${sourceImageId}`,
      { method: 'POST', signal },
    ),

  /** Endpoint 38 — `GET /images/{id}/gcps` → `200 Page[GcpRead]`. */
  listForImage: (imageId: Uuid, f: GcpFilters = {}, signal?: AbortSignal): Promise<Page<GcpRead>> =>
    fetchJson<Page<GcpRead>>(forImage(imageId), {
      query: { format: 'json', ...toQuery(f) },
      signal,
    }),

  /**
   * Endpoint 38 with `format=geojson` → a `GeoJsonFeatureCollection`.
   *
   * ★ `[lon, lat]` inside the geometry — RFC 7946 mandates x,y — while `lat`/`lon`
   *   objects everywhere else are lat-first (§6.1). The inversion is the single most
   *   common GeoJSON defect; it is stated here because this is the one function in
   *   the unit that returns it.
   */
  listForImageGeoJson: (
    imageId: Uuid,
    f: GcpFilters = {},
    signal?: AbortSignal,
  ): Promise<GeoJsonFeatureCollection<GcpRead>> =>
    fetchJson<GeoJsonFeatureCollection<GcpRead>>(forImage(imageId), {
      query: { format: 'geojson', ...toQuery(f) },
      signal,
    }),

  /** Endpoint 64 — `GET /match-results/{id}/gcps` → `200 Page[GcpRead]`. Empty in this build. */
  listForMatch: (
    matchResultId: Uuid,
    f: GcpFilters = {},
    signal?: AbortSignal,
  ): Promise<Page<GcpRead>> =>
    fetchJson<Page<GcpRead>>(`/match-results/${matchResultId}/gcps`, { query: toQuery(f), signal }),

  /**
   * Endpoint 39 — `GET /gcps/{id}` → `200 GcpRead` + `ETag`.
   *
   * ★ Reading a GCP is what puts its `ETag` in the registry, which is what makes the
   *   `If-Match` on {@link gcpsApi.update} possible. A list response carries one ETag
   *   for the whole page, so it cannot supply a per-row one — `useAdjustGcp`
   *   documents the consequence.
   */
  get: (id: Uuid, signal?: AbortSignal): Promise<GcpRead> =>
    fetchJson<GcpRead>(flat(id), { signal }),

  /**
   * ★ `GET /gcps` → `200 Page[GcpOverview]` — EVERY GCP THE SURVEYOR HAS PLACED,
   *   across every project, for the dashboard's geographic overview.
   *
   * ★ Unnested deliberately. Every other read here is nested under an image or a
   *   match result, because every other screen is *about* one photograph. The
   *   dashboard is about the whole body of work, and fanning out one request per
   *   project to assemble it would make the first screen of the app the slowest.
   *
   * ★ It returns {@link GcpOverview}, not `GcpRead` — see the type for why.
   */
  overview: (f: GcpOverviewFilters = {}, signal?: AbortSignal): Promise<Page<GcpOverview>> =>
    fetchJson<Page<GcpOverview>>('/gcps', { query: toQuery(f), signal }),

  /**
   * ★ Endpoint 40 — `PATCH /gcps/{id}`, **`If-Match` REQUIRED** → `200` · `412` · `422`.
   *
   *   THE ADJUSTMENT ENDPOINT (SCOPE.md §5: *"either endpoint can be dragged and
   *   re-committed"*). §8.5 is explicit that `POST /gcps/{id}/adjust` — 50-frontend's
   *   design, with `{imagePixel, latLon, pinned}` — is **VOID**: it does not exist,
   *   and `pinned` is on no schema and no table.
   *
   * ★ `image_px` IS NOT ADJUSTABLE HERE. Moving the point on the photograph is
   *   `PATCH /annotations/{id}`. Two endpoints, two meanings: *"the coordinate is in
   *   the wrong place"* vs *"I marked the wrong pixel"*. `GcpUpdate` has no `image_px`
   *   field, so the type system enforces it.
   */
  update: (id: Uuid, body: GcpUpdate, etag?: string, signal?: AbortSignal): Promise<GcpRead> =>
    fetchJson<GcpRead>(flat(id), { method: 'PATCH', body, ifMatch: etag, signal }),

  /**
   * Endpoint 41 — `POST /gcps/{id}/reset` → `200 GcpRead`. **Idempotent.**
   *
   * ★ Restores `original` — the algorithm's answer, written once and kept forever.
   *   `422 GCP_ORIGINAL_UNAVAILABLE` when the GCP was never adjusted.
   */
  reset: (
    id: Uuid,
    body: GcpResetRequest = { confirm: true },
    signal?: AbortSignal,
  ): Promise<GcpRead> => fetchJson<GcpRead>(`${flat(id)}/reset`, { method: 'POST', body, signal }),

  /**
   * ★★ ADDED — `DELETE /gcps/{id}` → `204`. Permanently removes a manual GCP.
   *
   * ★ A hard delete on the server (manual GCPs carry no `deleted_at`). The backing
   *   landmark annotation is left intact — removing a control point is not removing the
   *   feature it marked.
   */
  remove: async (id: Uuid, signal?: AbortSignal): Promise<void> => {
    await fetchJson<void>(flat(id), { method: 'DELETE', parse: 'none', signal });
    forgetEtag(flat(id));
  },

  /**
   * ★ DEFERRED — endpoint 42 `POST /images/{id}/gcps/recompute` → **`501`** in this
   *   build. Both modes are deferred: `rederive` runs the matching pipeline and
   *   `refit` re-fits a homography (SCOPE.md §4). Live: `202 JobRead`, or `200
   *   GcpRecomputeDryRun` when `dry_run: true`.
   */
  recompute: (
    imageId: Uuid,
    body: GcpRecomputeRequest = {},
    signal?: AbortSignal,
  ): Promise<GcpRecomputeResponse> =>
    fetchJson<GcpRecomputeResponse>(`${forImage(imageId)}/recompute`, {
      method: 'POST',
      body,
      signal,
    }),

  /** Drops a GCP's cached `ETag`. Called after any write that supersedes it. */
  forgetEtag: (id: Uuid): void => forgetEtag(flat(id)),
};

File: backend/app/api/v1/gcps.py
"""GCPs — endpoints 38–42, 64, plus the SCOPE.md §5 additions 65 and 66.

★★ **THE CORE OF THIS BUILD** (SCOPE.md §5). With automatic matching deferred, a GCP is a
**manual correspondence**: the surveyor marks a landmark in the photo and clicks the same spot
on the satellite map. The coordinate is a **direct observation**, ``source='manual'``.

* 65 ``POST /images/{id}/gcps``            — create a manual GCP (the product's core write).
* 66 ``PUT /gcps/{id}/correspondence``     — re-commit the pairing (either endpoint draggable).
* 40 ``PATCH /gcps/{id}``                  — adjust / metadata edit (``If-Match`` REQUIRED).
* 41 ``POST /gcps/{id}/reset``             — restore the original answer (idempotent-safe).
* 42 ``POST /images/{id}/gcps/recompute``  — **DEFERRED** (501): needs the homography estimator.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api import deps, presenters
from app.api.etag import etag_for, if_match_ok
from app.core.config import Settings
from app.api.v1.projects import parse_bbox
from app.core.exceptions import StaleGcpVersion
from app.core.pagination import SortKey, SortParams
from app.observability import metrics
from app.schemas.common import GeoJsonFeature, GeoJsonFeatureCollection, Page
from app.schemas.gcp import (
    GCP_SORT_FIELDS,
    GcpCorrespondenceUpdate,
    GcpCopyResult,
    GcpManualCreate,
    GcpOverview,
    GcpRead,
    GcpResetRequest,
    GcpUpdate,
)
from app.services.gcp_service import GcpService

router_image = APIRouter()
router_flat = APIRouter()
router_results = APIRouter()

_DEFAULT_SORT = (SortKey("created_at", descending=True),)


def _gcp_etag(gcp) -> str:
    return etag_for(int(gcp.updated_at.timestamp() * 1_000_000) if gcp.updated_at else 0, gcp.updated_at)


def _feature_collection(gcps) -> GeoJsonFeatureCollection:
    features: list[GeoJsonFeature] = []
    for gcp in gcps:
        read = presenters.to_gcp_read(gcp)
        features.append(
            GeoJsonFeature(
                geometry={"type": "Point", "coordinates": [read.lon, read.lat]},
                properties={
                    "id": str(read.id),
                    "code": read.code,
                    "confidence": read.confidence,
                    "source": read.source.value if hasattr(read.source, "value") else read.source,
                    "total_ce90_m": read.accuracy.total_ce90_m,
                },
                id=str(read.id),
            )
        )
    return GeoJsonFeatureCollection(features=features)


# ── router_image ────────────────────────────────────────────────────────────────


@router_image.get("/images/{image_id}/gcps", summary="List an image's GCPs")
async def list_gcps(
    format: str = Query("json", pattern="^(json|geojson)$"),
    match_result_id: UUID | None = Query(None),
    source: str | None = Query(None, description="manual|automatic (SCOPE.md §5 filter)."),
    confidence_gte: float | None = Query(None, alias="confidence__gte", ge=0.0, le=100.0),
    confidence_lte: float | None = Query(None, alias="confidence__lte", ge=0.0, le=100.0),
    manually_adjusted: bool | None = Query(None),
    is_stale: bool | None = Query(None),
    is_included_in_export: bool | None = Query(None),
    q: str | None = Query(None),
    pagination=Depends(deps.get_pagination),
    sort: str | None = Depends(deps.get_sort),
    image=Depends(deps.get_image),
    db: AsyncSession = Depends(deps.get_db),
):
    parsed = SortParams.parse(sort, allowed=GCP_SORT_FIELDS, default=_DEFAULT_SORT)
    rows, total = await deps.list_gcps_for_image(
        db,
        image.id,
        pagination=pagination,
        sort=parsed,
        match_result_id=match_result_id,
        source=source,
        confidence_gte=confidence_gte,
        confidence_lte=confidence_lte,
        manually_adjusted=manually_adjusted,
        is_stale=is_stale,
        is_included_in_export=is_included_in_export,
        q=q,
    )
    if format == "geojson":
        return _feature_collection(rows)
    return Page.of([presenters.to_gcp_read(g) for g in rows], total, pagination)


@router_image.post(
    "/images/{image_id}/gcps",
    response_model=GcpRead,
    status_code=status.HTTP_201_CREATED,
    summary="Create a manual GCP (SCOPE.md §5)",
)
async def create_manual_gcp(
    body: GcpManualCreate,
    request: Request,
    response: Response,
    image=Depends(deps.get_image),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> GcpRead:
    from app.services.imagery_service import ImageryService

    service = GcpService(
        db, settings, imagery=ImageryService(settings, registry=deps.get_imagery_registry(request))
    )
    gcp = await service.create_manual(image, body)
    await db.flush()
    response.headers["Location"] = f"{request.scope.get('root_path', '')}/api/v1/gcps/{gcp.id}"
    return presenters.to_gcp_read(gcp)


@router_image.post(
    "/images/{image_id}/gcps/copy-from/{source_image_id}",
    response_model=GcpCopyResult,
    status_code=status.HTTP_201_CREATED,
    summary="Carry another frame's control points onto this one (same camera, same aim)",
)
async def copy_gcps_from(
    source_image_id: UUID,
    request: Request,
    image=Depends(deps.get_image),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> GcpCopyResult:
    """★ A NEW FRAME, THE SAME AIM (2026-09-09, owner ask). When a camera adopts a
    fresh frame and the operator says the camera has not moved, every control
    point of the previous frame is valid at the same pixel — so it is recreated
    on the new frame through the ordinary manual-GCP path (its own annotation,
    its own elevation sample, provenance kept) instead of being placed again.
    Both frames must belong to the same project."""
    from app.db.repositories.images import ImageRepository
    from app.services.imagery_service import ImageryService

    source = await ImageRepository(db).get_active(source_image_id)
    if source is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"No image {source_image_id}."
        )
    if source.project_id != image.project_id:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="control points can only be carried between frames of the same project.",
        )
    service = GcpService(
        db, settings, imagery=ImageryService(settings, registry=deps.get_imagery_registry(request))
    )
    copied = await service.copy_to(source, image)
    await db.flush()
    return GcpCopyResult(copied=len(copied), items=[presenters.to_gcp_read(g) for g in copied])


@router_image.post("/images/{image_id}/gcps/recompute", summary="Recompute GCPs (DEFERRED)")
async def recompute_gcps(
    image=Depends(deps.get_image),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
):
    # ★ DEFERRED (SCOPE.md §4): both refit and rederive need the homography estimator. 501.
    GcpService(db, settings).recompute(image.id)


# ── router_flat ─────────────────────────────────────────────────────────────────


@router_flat.get(
    "/gcps",
    response_model=Page[GcpOverview],
    summary="List GCPs across every project (dashboard map)",
)
async def list_gcps_overview(
    project_id: UUID | None = Query(None),
    image_id: UUID | None = Query(None),
    album_id: UUID | None = Query(None, description="Every GCP in this album's projects."),
    min_confidence: float | None = Query(None, ge=0.0, le=100.0),
    bbox: str | None = Query(None, description='"minlon,minlat,maxlon,maxlat", EPSG:4326.'),
    pagination=Depends(deps.get_pagination),
    sort: str | None = Depends(deps.get_sort),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
) -> Page[GcpOverview]:
    """★ The **only** cross-project view of the deliverable.

    Newest first by default. ``project_name``, ``image_filename`` and ``landmark_name``
    are resolved by joins inside the repository — one statement for the whole page, no
    matter how many rows, because this feeds a map that may render thousands of markers.
    """
    parsed = SortParams.parse(sort, allowed=GCP_SORT_FIELDS, default=_DEFAULT_SORT)
    rows, total = await GcpService(db, settings).list_overview(
        pagination=pagination,
        sort=parsed,
        project_id=project_id,
        image_id=image_id,
        album_id=album_id,
        min_confidence=min_confidence,
        bbox=parse_bbox(bbox),
    )
    return Page.of([presenters.to_gcp_overview(r) for r in rows], total, pagination)


@router_flat.get("/gcps/{gcp_id}", response_model=GcpRead, summary="Get a GCP")
async def get_gcp(response: Response, gcp=Depends(deps.get_gcp)) -> GcpRead:
    response.headers["ETag"] = _gcp_etag(gcp)
    return presenters.to_gcp_read(gcp)


@router_flat.patch("/gcps/{gcp_id}", response_model=GcpRead, summary="Adjust a GCP")
async def update_gcp(
    body: GcpUpdate,
    request: Request,
    response: Response,
    gcp=Depends(deps.get_gcp),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> GcpRead:
    supplied = deps.require_if_match(request)  # ★ If-Match REQUIRED → 428 when absent
    if not if_match_ok(supplied, _gcp_etag(gcp)):
        raise StaleGcpVersion(f"GCP {gcp.id} was modified since your ETag; refetch and retry.")
    updated = await GcpService(db, settings).adjust(gcp, body)
    await db.flush()
    if updated.manually_adjusted:
        try:
            metrics.gcp_adjustments_total.inc()
        except Exception:  # noqa: BLE001
            pass
    response.headers["ETag"] = _gcp_etag(updated)
    return presenters.to_gcp_read(updated)


@router_flat.post("/gcps/{gcp_id}/reset", response_model=GcpRead, summary="Reset a GCP to its original")
async def reset_gcp(
    gcp_id: UUID,
    body: GcpResetRequest,
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> GcpRead:
    gcp = await GcpService(db, settings).reset(gcp_id)
    await db.flush()
    return presenters.to_gcp_read(gcp)


@router_flat.delete("/gcps/{gcp_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete a GCP")
async def delete_gcp(
    gcp=Depends(deps.get_gcp),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> Response:
    """Permanently remove a manual GCP.

    ★ A hard delete, by design: a manual GCP is the surveyor's own observation, not
    algorithm output that needs a tombstone for provenance — the ``gcps`` table carries no
    ``deleted_at``, nothing references a GCP as a FK child, and the backing landmark
    annotation is left intact (deleting a control point is not deleting the feature it
    marked). ``get_gcp`` already 404s an unknown id.
    """
    await GcpService(db, settings).delete(gcp)
    await db.flush()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router_flat.put("/gcps/{gcp_id}/correspondence", response_model=GcpRead, summary="Re-commit the pairing (SCOPE.md §5)")
async def update_correspondence(
    body: GcpCorrespondenceUpdate,
    request: Request,
    response: Response,
    gcp=Depends(deps.get_gcp),
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> GcpRead:
    supplied = deps.require_if_match(request)  # ★ If-Match REQUIRED
    if not if_match_ok(supplied, _gcp_etag(gcp)):
        raise StaleGcpVersion(f"GCP {gcp.id} was modified since your ETag; refetch and retry.")
    from app.services.imagery_service import ImageryService

    service = GcpService(
        db, settings, imagery=ImageryService(settings, registry=deps.get_imagery_registry(request))
    )
    updated = await service.update_correspondence(gcp, body)
    await db.flush()
    response.headers["ETag"] = _gcp_etag(updated)
    return presenters.to_gcp_read(updated)


# ── router_results ───────────────────────────────────────────────────────────────


@router_results.get("/match-results/{match_result_id}/gcps", response_model=Page[GcpRead], summary="A match result's GCPs")
async def list_gcps_for_match_result(
    match_result_id: UUID,
    pagination=Depends(deps.get_pagination),
    sort: str | None = Depends(deps.get_sort),
    db: AsyncSession = Depends(deps.get_db),
) -> Page[GcpRead]:
    parsed = SortParams.parse(sort, allowed=GCP_SORT_FIELDS, default=_DEFAULT_SORT)
    rows, total = await deps.list_gcps_for_match_result(
        db, match_result_id, pagination=pagination, sort=parsed
    )
    return Page.of([presenters.to_gcp_read(g) for g in rows], total, pagination)
